File: elixir/bracket-push/bracket_push.exs
defmodule BracketPush do
  @openers ["{", "[", "("]
  @closers ["}", "]", ")"]

  @doc """
  Checks that all the brackets and braces in the string are matched correctly, and nested correctly
  """
  @spec check_brackets(String.t) :: boolean
  def check_brackets(str) do str
    |> String.split(~r//, trim: true)
    |> balance([])
  end

  defp balance([], []), do: true
  defp balance([], _), do: false
  defp balance(["}"|_], [current|_]) when current != "{", do: false
  defp balance(["]"|_], [current|_]) when current != "[", do: false
  defp balance([")"|_], [current|_]) when current != "(", do: false
  defp balance([h|t], stack) when h in @openers, do: balance(t, [h] ++ stack)
  defp balance([h|t], [_|new_stack]) when h in @closers, do: balance(t, new_stack)
  defp balance([_|tail], stack), do: balance(tail, stack)
end

File: erlang/bracket-push/src/bracket_push.erl
-module(bracket_push).
-export([is_paired/1]).

is_paired(String) -> balance(String, []).

balance([], []) -> true;
balance([], _) -> false;
% An opener. Add it to the stack and continue.
balance([Head | Tail], Stack) when Head == $(; Head == $[; Head == ${ -> balance(Tail, [Head | Stack]);
% A closer that matches the opener on the top of the stack. Pop it from the stack and continue.
balance([$) | Tail], [$( | Rest]) -> balance(Tail, Rest);
balance([$] | Tail], [$[ | Rest]) -> balance(Tail, Rest);
balance([$} | Tail], [${ | Rest]) -> balance(Tail, Rest);
% A closer that does not match the opener on the top of the stack. Return false.
balance([Head | _], _) when Head == $); Head == $]; Head == $} -> false;
% Not an opener or closer. Discard it and continue.
balance([_ | Tail], Stack) -> balance(Tail, Stack).
